### ISSUE TITLE
"'CharmRelation' object is not subscriptable" errors

### ISSUE THREAD

**Issue by Vultaire** [juju/python-libjuju#414]
Hello,

At least one of our charms (likely others, but I've only tested one thus far) is no longer able to run libjuju-based tests without modification.

It seems that commit 4d2895a8737c1b6c89a8322c0a69c8ea2236c431 introduced changes where data which was formerly in dictionaries is now in CharmRelation objects (as seen in e.g. juju/client/_definitions.py, AddRelationResults class).  This seems to be breaking things internally to libjuju.

Here's an example of pytest output upon trying to call add_relation():

```
        for ubuntu_unit in ['ubuntu', 'ubuntu-checksum']:
            await nrpe_app.add_relation(
                'general-info',
>               '{}-{}:juju-info'.format(ubuntu_unit, release)
            )

tests/functional/test_hwhealth.py:94: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
/usr/lib/python3.6/asyncio/coroutines.py:110: in __next__
    return self.gen.send(None)
.tox/functional/lib/python3.6/site-packages/juju/application.py:103: in add_relation
    return await self.model.add_relation(local_relation, remote_relation)
/usr/lib/python3.6/asyncio/coroutines.py:110: in __next__
    return self.gen.send(None)
.tox/functional/lib/python3.6/site-packages/juju/model.py:1192: in add_relation
    for app, data in result.endpoints.items()]
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

.0 = <dict_itemiterator object at 0x7f76939210e8>

    specs = ['{}:{}'.format(app, data['name'])
>            for app, data in result.endpoints.items()]
E   TypeError: 'CharmRelation' object is not subscriptable

.tox/functional/lib/python3.6/site-packages/juju/model.py:1192: TypeError
```

**Comment by battlemidget** [comment]
We're hitting this in our CI as well, pretty much blocking all testing

**Comment by Vultaire** [comment]
Temporary workaround for the above is to use juju<2.8.0 in your requirements.txt.  I called out the commit but not the version; this is specific to 2.8.0.

**Comment by hpidcock** [comment]
@Vultaire can you please test with https://github.com/juju/python-libjuju/tarball/master#egg=juju-2.8.0

**Comment by hpidcock** [comment]
Fixed in #416
